### PR TITLE
[Messenger] Proper message decoding error handling

### LINKED ISSUE #44117: [Messenger] MessageDecodingFailedException should not delete message from queue
### Symfony version(s) affected

All

### Description

`MessageDecodingFailedException` should not delete a message from the queue.

```
try {
    $envelope = $this->serializer->decode([
        'body' => $sqsEnvelope['body'],
        'headers' => $sqsEnvelope['headers'],
    ]);
} catch (MessageDecodingFailedException $exception) {
    $this->connection->delete($sqsEnvelope['id']);

    throw $exception;
}
```

### Issues it causes

- Since decoding is not exactly reverse of encoding with arrays (int vs string keys), legitimate messages end up being dropped.
- Any other kind of dev ops error can mess up the decoding temporarily.

### PR DESCRIPTION
**Pull request by alex-dev**
| Q | A
|---|---
| Branch? | 6.3
| Bug fix? | yes
| New feature? | no
| Deprecations? | no
| BC breaks?  | yes
| Tickets | Fix #44117
| License | MIT
| Doc PR | N/A

- [ ] Update CHANGELOG.md
- [x] Update whatever is needed for BC breaks
- [x] Ensure serialization works across a full failing message (failed decoding into encoding)

## Problem
Handling deserialization in transports prevents actual error handling from running... And depending on transports, can even drop messages. While envelope and stamps deserialization failures may be actual error warranting undefined behavior, message denormalization failures is merely an application concern. Symfony should never drop messages that are broken by the application.

## Solution
Reusing some part of #39622:
- Ensure both `PhpSerializer` and `Serializer` do not throw error in message deserialization
  - Stamps and envelope are still allowed to throw.
  - Remove toggling incomplete classes from `PhpSerializer`
- Handle `MessageDecodingFailedStamp` in `Worker`
  - Trigger standard message rejection (event and retry into full rejection)
- Ensure compatibility with `AbstractFailedMessagesCommand`

## BC breaks
- `Messenger\Transport\Serialization\Serializer::__construct()` first arguments goes from `?SerializerInterface $serializer = null` to `DecoderInterface&DenormalizerInterface&SerializerInterface $serializer`
  - To support distinguishing between encoding errors and normalization errors, we must call each separately. This requires an intersection of interfaces
  - Framework always pass a constructed serializer, so no issue
  - No other instantiations are done... So only BC breaks will be in standalone uses passing null as first argument
  - `Messenger\Transport\Serialization\Serializer::create()` still return a default constructed instance

### CODE REVIEW

File: src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php [hunk 1]
```diff
@@ -69,20 +70,24 @@ public function decode(array $encodedEnvelope): Envelope
...
         }
 
         try {
-            $message = $this->serializer->deserialize($encodedEnvelope['body'], $encodedEnvelope['headers']['type'], $this->format, $context);
-        } catch (ExceptionInterface $e) {
-            throw new MessageDecodingFailedException('Could not decode message: '.$e->getMessage(), $e->getCode(), $e);
+            $message = $this->serializer->denormalize($body, $type, $this->format, $context);
+        } catch (\Throwable $e) {
+            $stamps[] = new MessageDecodingFailedStamp($e->getMessage());
+            $message = $this->serializer->denormalize($body, \stdClass::class, $this->format, $context);
```
Comment (nikophil) [inline src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php:90]: not sure it is very common, but what happens if a handler consumes messages of class `\stdClass::class`? or if `allow_no_handlers` is activated?

FMI, why do we need to separate the two behaviors, when decoding fails and when denormalization fails?

Comment (alex-dev) [reply src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php:90]: - Anything related to handlers does not matter here. We handle the denormalization failure using stamps long before handlers.
- We split decoding and denormalization to match `PhpSerializer` behavior when it allows incomplete classes: an `Envelope` must be generated, message must be a valid object.

**Comment by alex-dev** [comment]
@lyrixx Here is a fix for #44117

**Comment by ismail1432** [comment]
Hello @alex-dev,
I think this fix is important, why the PR was closed?

**Comment by alex-dev** [comment]
I could not justify spending more time on this to update it. Seeing Symfony had  little interest in this fix, my company chose to implement a solution inside the transport layer. It is hacked, but the right behavior is guaranteed.
If thee is clear interest in this, I may reopen in January.

**Comment by B-Galati** [comment]
@alex-dev any chance you could share your solution?

**Comment by alex-dev** [comment]
We just reimplemented the transport and made rejection a no-op. There are a few annoying caveats. 
- Retries and failures must be handled by the queue itself (SQS does).
- You lose a lot of debugging utilities Symfony gives you (you must unconfigure failure handling from Symfony)

Other alternative would be to reimplement the worker and associated commands (what this PR more or less does).

**Pull request closed without merging**